transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java

```diff
@@ -0,0 +1,218 @@
+  private final long arrayAddress;
+
+  private final long submissionQueueArrayAddress;
+
+  private long sqeHead;
+  private long sqeTail;
+
+  private final int ringSize;
+  private final long ringAddress;
+  private final int ringFd;
+
+  public IOUringSubmissionQueue(long kHeadAddress, long kTailAddress, long kRingMaskAddress, long kRingEntriesAddress,
+      long fFlagsAdress, long kDroppedAddress, long arrayAddress, long submissionQueueArrayAddress, int ringSize,
+      long ringAddress, int ringFd) {
+    this.kHeadAddress = kHeadAddress;
+    this.kTailAddress = kTailAddress;
+    this.kRingMaskAddress = kRingMaskAddress;
+    this.kRingEntriesAddress = kRingEntriesAddress;
+    this.fFlagsAdress = fFlagsAdress;
+    this.kDroppedAddress = kDroppedAddress;
+    this.arrayAddress = arrayAddress;
+    this.submissionQueueArrayAddress = submissionQueueArrayAddress;
+    this.ringSize = ringSize;
+    this.ringAddress = ringAddress;
+    this.ringFd = ringFd;
+  }
+
+  public long getSqe() {
+    long next = sqeTail + 1;
+    long kRingEntries = toUnsignedLong(PlatformDependent.getInt(kRingEntriesAddress));
+    long sqe = 0;
+    if ((next - sqeHead) <= kRingEntries) {
+      long index = sqeTail & toUnsignedLong(PlatformDependent.getInt(kRingMaskAddress));
+      sqe = SQE_SIZE * index + submissionQueueArrayAddress;
+      sqeTail = next;
+    }
+    return sqe;
+  }
+
+  private void setData(long sqe, long eventId, EventType type, int fd, long bufferAddress, int length, long offset) {
+    //Todo cleaner
+    PlatformDependent.putByte(sqe + SQE_OP_CODE_FIELD, (byte) type.getOp());
+    PlatformDependent.putByte(sqe + SQE_FLAGS_FIELD, (byte) 0);
+    PlatformDependent.putShort(sqe + SQE_IOPRIO_FIELD, (short) 0);
+    PlatformDependent.putInt(sqe + SQE_FD_FIELD, fd);
+    PlatformDependent.putLong(sqe + SQE_OFFSET_FIELD, offset);
+    PlatformDependent.putLong(sqe + SQE_ADDRESS_FIELD, bufferAddress);
+    PlatformDependent.putInt(sqe + SQE_LEN_FIELD, length);
+    PlatformDependent.putInt(sqe + SQE_RW_FLAGS_FIELD, 0);
+    PlatformDependent.putLong(sqe + SQE_USER_DATA_FIELD, eventId);
+
+    // pad field array -> all fields should be zero
+    long offsetIndex = 0;
+    for (int i = 0; i < 3; i++) {
+      PlatformDependent.putLong(sqe + SQE_PAD_FIELD + offsetIndex, 0);
+      offsetIndex += 8;
+    }
+
+    System.out.println("OPField: " + PlatformDependent.getByte(sqe + SQE_OP_CODE_FIELD));
+    System.out.println("UserDataField: " + PlatformDependent.getByte(sqe + SQE_USER_DATA_FIELD));
+  }
+
+  public boolean add(long eventId, EventType type, int fd, long bufferAddress, int pos, int limit) {
+    long sqe = getSqe();
+    if (sqe == 0) {
+      return false;
+    }
+    setData(sqe, eventId, type, fd, bufferAddress + pos, limit - pos, 0);
+    return true;
+  }
+
+  private int flushSqe() {
+    long kTail = toUnsignedLong(PlatformDependent.getInt(kTailAddress));
+    long kHead = toUnsignedLong(PlatformDependent.getInt(kHeadAddress));
```

Comment:
For the SQE ring, I think we need to read the `head` value with `getIntVolatile`.

Reply:
I don't think there will compiler memory reordering because of that https://github.com/netty/netty/blob/692238f6da4beb8e6698e214bfadc7d1135225d6/transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java#L125
what do you think? @chrisvest

Reply:
The `head` field is updated by the kernel. It's the memory visibility semantics I'm after in this case.